pkg/types/hashedrekord/v0.0.1/entry.go

```diff
@@ -0,0 +1,264 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package hashedrekord
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"strings"
+
+	"github.com/asaskevich/govalidator"
+	"github.com/go-openapi/strfmt"
+	"github.com/go-openapi/swag"
+	"github.com/pkg/errors"
+
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/sigstore/rekor/pkg/log"
+	"github.com/sigstore/rekor/pkg/pki"
+	"github.com/sigstore/rekor/pkg/types"
+	hashedrekord "github.com/sigstore/rekor/pkg/types/hashedrekord"
+	"github.com/sigstore/sigstore/pkg/signature/options"
+)
+
+const (
+	APIVERSION = "0.0.1"
+)
+
+func init() {
+	if err := hashedrekord.VersionMap.SetEntryFactory(APIVERSION, NewEntry); err != nil {
+		log.Logger.Panic(err)
+	}
+}
+
+type V001Entry struct {
+	HashedRekordObj models.HashedrekordV001Schema
+	keyObj          pki.PublicKey
+	sigObj          pki.Signature
+}
+
+func (v V001Entry) APIVersion() string {
+	return APIVERSION
+}
+
+func NewEntry() types.EntryImpl {
+	return &V001Entry{}
+}
+
+func (v V001Entry) IndexKeys() []string {
+	var result []string
+
+	key, err := v.keyObj.CanonicalValue()
+	if err != nil {
+		log.Logger.Error(err)
+	} else {
+		keyHash := sha256.Sum256(key)
+		result = append(result, strings.ToLower(hex.EncodeToString(keyHash[:])))
+	}
+
+	result = append(result, v.keyObj.EmailAddresses()...)
+
+	if v.HashedRekordObj.Data.Hash != nil {
+		hashKey := strings.ToLower(fmt.Sprintf("%s:%s", *v.HashedRekordObj.Data.Hash.Algorithm, *v.HashedRekordObj.Data.Hash.Value))
+		result = append(result, hashKey)
+	}
+
+	return result
+}
+
+func (v *V001Entry) Unmarshal(pe models.ProposedEntry) error {
+	rekord, ok := pe.(*models.Hashedrekord)
+	if !ok {
+		return errors.New("cannot unmarshal non Rekord v0.0.1 type")
+	}
+
+	if err := types.DecodeEntry(rekord.Spec, &v.HashedRekordObj); err != nil {
+		return err
+	}
+
+	// field validation
+	if err := v.HashedRekordObj.Validate(strfmt.Default); err != nil {
+		return err
+	}
+
+	// cross field validation
+	return v.validate()
+}
+
+func (v *V001Entry) Canonicalize(ctx context.Context) ([]byte, error) {
+	if err := v.validate(); err != nil {
+		return nil, types.ValidationError(err)
+	}
+
+	artifactFactory, err := pki.NewArtifactFactory(pki.X509)
+	if err != nil {
+		return nil, err
+	}
+
+	signature, err := artifactFactory.NewSignature(bytes.NewReader(v.HashedRekordObj.Signature.Content))
+	if err != nil {
+		return nil, err
+	}
+	v.sigObj = signature
+
+	key, err := artifactFactory.NewPublicKey(bytes.NewReader(v.HashedRekordObj.Signature.PublicKey.Content))
+	if err != nil {
+		return nil, err
+	}
+	v.keyObj = key
```

Comment:
is there a reason to reparse and overwriting `v.sigObj` and `v.keyObj` here? I think you should ensure they're not nil but otherwise they've already been parsed and loaded by the time `Canonicalize` has been called - so i don't think this is needed.

Reply:
Yep, fixed -- In most other types they're set by `FetchExternalEntities`, but here they're set during `validate()` during signature verification. Given that i had to make `validate()` modify the entry